src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -126,16 +128,46 @@ internal sealed class AddDocumentsAction(
             ImmutableArray<DocumentState> documents)
             : TranslationAction(oldProjectState, newProjectState)
         {
+            /// <summary>
+            /// Amount to break batches of documents into.  That allows us to process things in parallel, without also
+            /// creating too many individual actions that then need to be processed.
+            /// </summary>
+            public const int AddDocumentsBatchSize = 32;
+
             public readonly ImmutableArray<DocumentState> Documents = documents;
 
             public override async Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
             {
-                // Parse all the documents in parallel.
-                using var _ = ArrayBuilder<Task<SyntaxTree>>.GetInstance(this.Documents.Length, out var tasks);
+#if NETSTANDARD
+                using var _1 = ArrayBuilder<Task>.GetInstance(this.Documents.Length, out var tasks);
+
+                // We want to parse in parallel.  But we don't want to have too many parses going on at the same time.
+                // So we use a semaphore here to only allow that many in at a time.  Once we hit that amount, it will
+                // block further parallel work.  However, as the semaphore is released, new work will be let in.
+                var semaphore = new SemaphoreSlim(initialCount: AddDocumentsBatchSize);
+                foreach (var document in this.Documents)
+                {
+                    tasks.Add(Task.Run(async () =>
```

Comment:
>tasks.Add(Task.Run(async () => [](http://example.com/codeflow?start=20&length=30)

The only slight disadvantage of the current form that I see is that all the tasks are always created now and that cancellation will have to go through all of them. By having the semaphore usage outside the task, could that be avoided?